app/src/main/java/com/example/jrdeveloper/MainActivity.kt

```diff
@@ -2,46 +2,129 @@ package com.example.jrdeveloper
 
 import android.os.Bundle
 import androidx.activity.ComponentActivity
-import androidx.activity.compose.setContent
-import androidx.activity.enableEdgeToEdge
-import androidx.compose.foundation.layout.fillMaxSize
-import androidx.compose.foundation.layout.padding
-import androidx.compose.material3.Scaffold
-import androidx.compose.material3.Text
-import androidx.compose.runtime.Composable
-import androidx.compose.ui.Modifier
-import androidx.compose.ui.tooling.preview.Preview
-import com.example.jrdeveloper.ui.theme.JrDeveloperTheme
+import android.annotation.SuppressLint
+import android.os.PersistableBundle
+import android.util.Log
+import android.widget.Button
+import android.widget.EditText
+import android.widget.TextView
+import android.widget.Toast
 
 class MainActivity : ComponentActivity() {
+
+    private lateinit var bank: BankAccount
```

Comment:
You should use lateinit when a variable needs to be initialized after the object is created but before it is used, and when you’re sure it will always be initialized before any access to it. However, in this case, bank is immediately reassigned inside onCreate. Since you already know the value for bank during initialization, you can directly initialize it there instead of declaring it as lateinit. This avoids the need for lateinit and simplifies the code. 

The bank variable declared as lateinit var at the class level is a global variable, but inside onCreate, a new local bank variable is created with the same name. This local bank variable shadows the global one, meaning any changes or initialization done to the local bank inside onCreate do not affect the global bank.

Later, when onSaveInstanceState tries to access the global bank, it finds that it was never initialized (since the lateinit variable was never assigned a value). This results in a crash with an UninitializedPropertyAccessException

Reply:
I commented out: private lateinit var bank: BankAccount
I'd imagine this code directly initializes it:
        val bank = BankAccount(
            accountNumber = "1234",
            accountHolder = "John Smith",
            balance = initialBalance
        )

Reply:
Ah, I see what you are saying about the local bank variable shadowing the global one.